clients/src/main/java/org/apache/kafka/common/metrics/stats/Meter.java

```diff
@@ -77,6 +77,8 @@ public Meter(TimeUnit unit, SampledStat rateStat, MetricName rateMetricName, Met
     @Override
     public void record(MetricConfig config, double value, long timeMs) {
         rate.record(config, value, timeMs);
-        total.record(config, value, timeMs);
+        // Total metrics with Count stat should record 1.0 (as recorded in the count)
+        double totalValue = (rate.stat instanceof Count) ? 1.0 : value;
```

Comment:
Not sure if this is the best fix. It seems that the issue is that the customized rateStat is only reflected in Rate, but not in total. It seems that we can extend Total to also allow a customizable stat. That way, rate and total can be updated consistently and we don't need customized code in record().

Reply:
@junrao Thanks for the review. I was trying to avoid public API changes and keep the change localized. Perhaps I missed something. Ideally, we would have a single stat used for both rate and total so they are consistent. But I couldn't figure out a way to make that work with the current APIs. The customized rate stat needs to be a SampledStat (e.g. Count), but the total stat needs to be cumulative. So don't we need two different, but consistent customized stats for rate and total? And caller needs to remember to make them consistent?

Reply:
Hey @rajinisivaram,

I've been wrestling with the same issue in the Streams module. The approach I took was actually to split up all my meters into separate rate and count/total metrics. The rationale being that we already have the ability to group a bunch of metrics under a sensor, why also have compound metrics? This allowed us to sidestep this issue, that sometimes you want the total over all values and other times you want to count the number of invocations. Rather than trying to infer the intent from the rate stat, we just explicitly add both metrics independently.

Of course, I failed to notice that Count is sampled, so now we have https://github.com/apache/kafka/pull/5467 to fix our total metrics again. But that's sort of beside the point.

Another unsolicited idea: it looks to me like there are only two kinds of meters in use: ones that use a `Count` rate and ones that use a `SampledTotal` rate. We could add two static factory methods, one for each usage, and privatize the other constructors. Of course, this is a public interface change, which you're trying to avoid...

Reply:
@junrao @vvcephei Yes, I totally agree that it would have been better to only allow value based Meter from start. But since we shouldn't change the public API now, I have added the check that @junrao suggested, so we only allow SampledTotal and Count.